Authentication: Replace encryption with hashing and add permission-based authorization

avirajsingh7 opened the pull request:
# Summary

This PR refactors the API key authentication system to improve security by replacing encryption-based key storage with hash-based verification. The new implementation uses a prefix-based lookup strategy combined with secure hashing, while maintaining backward compatibility with existing encrypted keys.

## Key Changes

### Security Improvements
- **Hash-based storage**: API keys are now hashed instead of encrypted, preventing key retrieval and enhancing security
- **Prefix-based lookup**: Keys consist of a prefix (for database lookup) and secret part (for verification)
- **Backward compatibility**: Support for verifying old encrypted API keys during transition period

### Architecture Updates
- **APIKeyManager class**: Replaced `verify_api_key` function with a comprehensive `APIKeyManager` class for better code organization
- **Permission system**: Added [permissions.py](backend/app/api/permissions.py) to separate authorization logic from authentication
- **Dependency refactoring**: Updated [deps.py](backend/app/api/deps.py) to focus solely on authentication, with permission checks delegated to the new permission module

### Database & Migration
- **Migration script**: Added [migrate_api_key.py](backend/app/alembic/migrate_api_key.py) to convert existing encrypted keys to hashed format and use uuid as primary key instead of int in APIkey table
- **Schema updates**: Updated API key model and database schema to support the new authentication logic
- **Delete endpoint**: Added API key deletion route with proper permission checks

### Testing Infrastructure
- **AuthContext pattern**: Refactored test cases to use `AuthContext` from seed data instead of generating keys (since hashed keys cannot be retrieved)
- **Seed data integration**: Updated test utilities to leverage [seed_data.json](backend/app/seed_data/seed_data.json) for consistent API key testing
- **Test coverage**: Updated all affected test suites across collections, documents, assistants, credentials, and conversations


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Create API keys (raw key shown once), list (masked) and delete per project with pagination
  * Unified auth context supporting API keys or JWTs and permission-based route guards

* **Improvements**
  * API keys migrated to a stronger hashed format with stable prefixes and UUID identifiers for improved security
  * Centralized API key manager and clearer error responses; raw key disclosure only at creation

* **Tests**
  * Expanded tests and test utilities covering auth, permissions, API key flows, migration, and verification
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

avirajsingh7 committed:
Draft one: modify model, added crud, route and fix deps, migration

avirajsingh7 committed:
pre commit

avirajsingh7 committed:
Add permission management for API key routes and user context dependency

avirajsingh7 committed:
Refactor API key management: replace verify_api_key function with APIKeyManager class, add delete API key route, and update related CRUD operations

avirajsingh7 committed:
Enhance API key management: add support for old key format in APIKeyManager, update key prefix extraction logic, and adjust seed data creation for API keys.

avirajsingh7 committed:
Fix API key generation: ensure exact lengths for prefix and secret key, and update comments for clarity.

avirajsingh7 committed:
Add migration script for API keys: convert encrypted keys to hashed format and update database schema

avirajsingh7 committed:
precommit

avirajsingh7 committed:
Refactor API key handling in tests to use AuthContext

- Removed direct usage of APIKeyPublic in tests and replaced with AuthContext for better encapsulation of authentication details.
- Updated test cases across various modules (e.g., test_api_key, test_assistants, test_creds, etc.) to utilize the new AuthContext structure.
- Simplified the retrieval of API keys and user context in test setups, enhancing readability and maintainability.
- Removed redundant API key creation logic from test utilities and centralized it within the AuthContext.
- Ensured all tests still pass with the new structure, maintaining functionality while improving code organization.

avirajsingh7 committed:
Fix API key route prefix and improve formatting in auth context functions

avirajsingh7 committed:
move APIKeyManager to security module and clean up imports in CRUD operations

avirajsingh7 committed:
user context handling to use AuthContext, updating dependencies and verification logic for API key authentication

avirajsingh7 committed:
rename AuthContext to TestAuthContext for consistency in test files

avirajsingh7 committed:
API key creation to include user and project ID parameters, enhancing user-specific key management.

avirajsingh7 committed:
enhance test coverage for API key CRUD and Routes operations

avirajsingh7 committed:
Add tests for API key manager

avirajsingh7 committed:
authentication context functions for consistency in test files

avirajsingh7 committed:
replace APIKeyCrud with create_test_api_key for consistency in test setup

avirajsingh7 committed:
Add tests for get_user_context

avirajsingh7 committed:
Add tests for permission checks and permission enum functionality

avirajsingh7 committed:
pre commt

avirajsingh7 committed:
Fix read_one method docstring and update logging for API key creation

avirajsingh7 committed:
Update API key deletion to set updated_at timestamp and modify user_id field in AuthContext

avirajsingh7 committed:
Fix import statement for get_project_by_id in APIKeyCrud

avirajsingh7 commented on backend/app/tests/utils/auth.py:
```diff
@@ -0,0 +1,146 @@
+        session=session,
+        user_email=settings.FIRST_SUPERUSER,
+        project_name="Glific",
+        raw_key="ApiKey No3x47A5qoIGhm0kVKjQ77dhCqEdWRIQZlEPzzzh7i8",
```
I will be looking into this,
how we can avoid hard coding here

nishika26 replied:
you can fetch it from seeded db? i dont know since i dont have full contecxt but still adding my input here

avirajsingh7 replied:
open 408 issue for this

avirajsingh7 committed:
Merge remote-tracking branch 'origin/main' into refactor/authentication

avirajsingh7 committed:
pre commit

avirajsingh7 committed:
fix migration

avirajsingh7 committed:
precommit

kartpop commented on backend/app/alembic/versions/e7c68e43ce6f_refactor_api_key_table.py:
```diff
@@ -0,0 +1,64 @@
+    # ### end Alembic commands ###
+
+
+def downgrade():
```
should there be something over here which does the reverse of `migrate_api_keys`?

what would be the repurcussions if we had to rollback and did nothing on downgrade?

avirajsingh7 replied:
instead of migration we will use db snapshot for downgrade.
Downgrade through script not possible due to hashing.
Added a comment.

kartpop commented on backend/app/core/security.py:
```diff
@@ -179,3 +186,145 @@ def decrypt_credentials(encrypted_credentials: str) -> dict:
+class APIKeyManager:
+    """
+    Handles API key generation and verification using secure hashing.
+    Supports Backwards compatibility with old key format.
```
> Supports Backwards compatibility with old key format.

we should expand on this i think in the doc string; add more details on old key lenght, new key length, what will change etc.

avirajsingh7 replied:
done

kartpop reviewed (commented):
WIP PR review - some initial comments

kartpop commented on backend/app/models/auth.py:
```diff
@@ -10,3 +13,13 @@ class Token(SQLModel):
+
+    user: User
+    organization: Organization | None = None
+    project: Project | None = None
```
any reason to have redundant fields here? 

if `user.id == user_id`, why have both `user` and `user_id` fields?

similar questions for project and organization

avirajsingh7 replied:
Since we’re already fetching the user, organization and project when validating the API key, we could just pass them along to the route instead of fetching again.
Reduce redundant lookup.

Well not sure how much this optimize, 
What's your opinion we can remove this also.

kartpop commented on backend/app/api/deps.py:
```diff
@@ -48,7 +48,7 @@ def get_current_user(
     """Authenticate user via API Key first, fallback to JWT token. Returns only User."""
 
     if api_key:
```
should be

```
if api_key:
    // logic
elif token:
    // logic
```

this is not part of the current change, but now that we have hit this part of the code, lets clean it up; i know we are returning from all paths inside the first if block, so logically it would not make a difference, but making this change would result in a cleaner, defensive code

avirajsingh7 replied:
Done.

kartpop commented on backend/app/api/routes/api_keys.py:
```diff
@@ -1,116 +1,81 @@
-    api_key = create_api_key(
-        session,
-        organization_id=project.organization_id,
+    The raw API key is returned only once during creation.
+    Store it securely as it cannot be retrieved again.
```
>    The raw API key is returned only once during creation.
>    Store it securely as it cannot be retrieved again.

i guess this message should also be sent in the `metadata` of the response; this is important info that should be known to the consumer of this endpoint

right now it is only superuser who can create this, so might be less relevant in the current scheme of things, but given that we already have a `metadata` field in our response payload, including this should be straightforward

avirajsingh7 replied:
added a message.

kartpop commented on backend/app/api/routes/api_keys.py:
```diff
@@ -1,116 +1,81 @@
-    Retrieve an API key by ID.
+    List all API keys for the current project.
+
+    Returns masked keys for security - the full key is only shown during creation.
```
where is the masking happening? 

also we are probably exposing our database scheme for `api_keys` as we use the same `APIKey` model for both create and read, this may not bee too bad given that we store only the prefix + hash. 

but why give out unnecessary info? why show the hash even? lets just mask it actually after the prefix part with a string of this sort `****`; better yet, the prefix also can be shown only partially, instead of 22 characters, show only first 15 or something and then `****`

the payload for this could be something like --> key_id, masked_key, inserted_at; not sure if anything else is required

```
    def read_all(self, skip: int = 0, limit: int = 100) -> list[APIKey]:
        """
        Read all non-deleted API keys for the project.
        """
        statement = (
            select(APIKey)
            .where(
                and_(
                    APIKey.project_id == self.project_id,
                    APIKey.is_deleted.is_(False),
                )
            )
            .offset(skip)
            .limit(limit)
        )
        return self.session.exec(statement).all()
```

avirajsingh7 replied:
There seems to be some confusion here.
We don’t return the key hash at any point. The API only returns the APIKeyPublic model, which excludes sensitive fields like the hashed secret.

When calling the read_all method, it fetches the full API key records from the database, including the hash. However, FastAPI automatically serializes the response according to the specified response_model — in this case, APIResponse[list[APIKeyPublic]].


```
@router.get(
    "/",
    response_model=APIResponse[list[APIKeyPublic]],
    dependencies=[Depends(require_permission(Permission.REQUIRE_PROJECT))],
)
```

Modified masked_key to key_prefix

nishika26 commented on backend/app/alembic/versions/e7c68e43ce6f_refactor_api_key_table.py:
```diff
@@ -0,0 +1,64 @@
+
+
+def downgrade():
+    pass
```
why is the downgrade here empty

avirajsingh7 replied:
refer [here](https://github.com/ProjectTech4DevAI/ai-platform/pull/396#discussion_r2434759085)

nishika26 commented on backend/app/api/permissions.py:
```diff
@@ -0,0 +1,70 @@
+from app.api.deps import AuthContextDep, SessionDep
+
+
+class Permission(str, Enum):
```
where is this permisson being set for the check

avirajsingh7 replied:
Each routes set a permission.
Check api_key routes.

nishika26 commented on backend/app/seed_data/seed_data.py:
```diff
@@ -49,7 +49,6 @@ class APIKeyData(BaseModel):
-    created_at: Optional[str] = None
 
 
 class CredentialData(BaseModel):
```
possible to confuse this with our byok credential module, maybe should be named AuthData

avirajsingh7 replied:
thats creds data only.

nishika26 commented on backend/app/models/auth.py:
```diff
@@ -10,3 +13,13 @@ class Token(SQLModel):
+    project_id: int | None = None
+    organization_id: int | None = None
+
+    user: User
```
what are these fields exactly

avirajsingh7 replied:
this is User Project Organization object.
more [context](https://github.com/ProjectTech4DevAI/ai-platform/pull/396#discussion_r2434767507)

nishika26 commented on backend/app/api/routes/api_keys.py:
```diff
@@ -1,116 +1,81 @@
+    response_model=APIResponse[Message],
+    dependencies=[Depends(require_permission(Permission.REQUIRE_PROJECT))],
+)
+def delete_api_key_route(
```
understandable that redundant logging would not be needed but there is no info log in crud delete as well as delete router, i feel like there should be one at least at one place, tracebility should be there for happy instances as well

avirajsingh7 replied:
there is already a log in delete method in crud.

nishika26 commented on backend/app/tests/core/test_security.py:
```diff
@@ -119,3 +124,185 @@ def test_get_encryption_key():
+
+    def test_verify_deleted_key(self, db: Session):
+        """Test that deleted API keys cannot be verified."""
+        # Create test API key
```
remove all these redundant comments

avirajsingh7 replied:
removed.

avirajsingh7 committed:
Refactor API key handling and improve documentation

- Updated downgrade function to clarify snapshot restoration.
- Changed conditional logic in get_current_user function for clarity.
- Enhanced create_api_key_route response with security message about raw API key.
- Updated APIKeyManager docstring
- removed unnecessary comments

avirajsingh7 committed:
Merge remote-tracking branch 'origin/main' into refactor/authentication

avirajsingh7 committed:
Update APIKeyManager docstring and enhance seed data comment for clarity

avirajsingh7 committed:
fix migration head

avirajsingh7 committed:
precomit

avirajsingh7 committed:
Refactor API key migration to use context manager for session handling and improve code clarity

avirajsingh7 committed:
precommit

avirajsingh7 committed:
remove id's from authcontext

avirajsingh7 committed:
Remove user_id from AuthContext instantiation in get_auth_context function

avirajsingh7 committed:
Update user authentication to check for active status before returning user object

avirajsingh7 committed:
precommit
